### PR TITLE
Add Spake2+ Implementation

### LINKED ISSUE #258: Missing openSSL based SPAKE2+ implementation
#### Problem
Missing SPAKE2+ implelementation
 
#### Proposed Solution
Implement it
@balducci-apple - any specific requirements here?

### PR DESCRIPTION
**Pull request by balducci-apple**
#### Problem
CHIP requires Spake2+ as a pairing mechanism.

 #### Summary of Changes

I've added a base class `Spake2p` which contains the necessary functions to run the Spake2+ protocol. This base class also acts as a PAL to some of the underlying cryptographic functions necessary for Spake2+ (point arithmetic, field arithmetic, etc). This base class, as implemented in CHIPCryptoPAL.cpp, is therefore independent of the underlying cryptographic library.

In addition, I've added `Spake2p_P256_SHA256_HKDF_HMAC` which represents the Spake2+ configuration to be used in CHIP 1.0. An implementation of the underlying cryptographic functions using OpenSSL is provided.

Finally, tests for the entire protocol as well as the underlying cryptographic functions are provided.
fixes #258

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Awesome, thanks @balducci-apple & @pan-apple !

**Comment by woody-apple** [comment]
Can we handle the LGTM issues in this PR?

**Comment by woody-apple** [comment]
@jelderton @saurabhst @robszewczyk ?

**Review by saurabhst — APPROVED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 1]
```diff
@@ -778,5 +780,287 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
     memset(buf, 0, len);
 }
 
+#define init_point(_point_)                                                                                                        \
+    _point_ = EC_POINT_new(context.curve);                                                                                         \
```
[the remark this answers is not in the document]
Comment (pan-apple) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp]: done

**Review by kedars — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton @hawk248 @robszewczyk ?

File: src/crypto/CHIPCryptoPAL.h [hunk 2]
```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
...
+     * @brief Return the shared secret.
+     *
+     * @param out     The output secret.
+     * @param out_len The output secret length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR GetKeys(unsigned char * out, size_t * out_len);
+
+    CHIP_SPAKE2P_ROLE role;
```
Comment (andy31415) [inline src/crypto/CHIPCryptoPAL.h:382]: do these members need to be public or could we make them a bit more restricted?

File: src/crypto/CHIPCryptoPAL.h [hunk 3]
```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
...
+     *     2. Initialize each of the field elements below and set the relevant pointers on the class:
+     *        a. w0
+     *        b. w1
+     *        c. xy
+     *        d. tempbn
+     *     3. The hashing context should be initialized
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR InitImpl() = 0;
```
Comment (andy31415) [inline src/crypto/CHIPCryptoPAL.h:423]: sholuld impl methods be prodtected?

File: src/crypto/CHIPCryptoPAL.h [hunk 4]
```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
...
+    /**
+     * @brief Write a point in 0x04 || X || Y format
+     *
+     * @param R       A pointer to an initialized implementation dependant point.
+     * @param out     Output buffer
+     * @param out_len Length of the output buffer
+     *
+     *  @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR PointWrite(const void * R, unsigned char * out, size_t out_len) = 0;
```
Comment (andy31415) [inline src/crypto/CHIPCryptoPAL.h:559]: nit: maybe having a typedef for the void* would increase readability. Thinking of:

PointWrite(const Point R, unsinged char * out, size_t out_len) 

as being faster to digest.

**Review by andy31415 — APPROVED** [review]
Approving however not submitting yet, please let me know if you would like to do some updates in this PR or iterate over subsequent PRs on my review points.

**Comment by gerickson** [comment]
@emargolis, @dougsteed, @roydsouza

**Comment by woody-apple** [comment]
Shoot, I hit merge on the wrong thing. @pan-apple can we address @andy31415 ’s comments?

**Comment by pan-apple** [comment]
> Approving however not submitting yet, please let me know if you would like to do some updates in this PR or iterate over subsequent PRs on my review points.

@andy31415 , I've taken note of your review comments, and we'll address these in a subsequent PR, as we still need to make changes WRT `mbedTLS` implementation of SPAKE2P.

File: src/crypto/CHIPCryptoPAL.h (hunk 2 above)
Comment (dougsteed) [reply src/crypto/CHIPCryptoPAL.h:382]: Also, shouldn't the data fields be private?

File: src/crypto/CHIPCryptoPAL.h [hunk 5]
```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
...
+    EVP_MD_CTX * hash_ctx;
+    const EVP_MD * hash;
+};
+#endif
+
+class Spake2p_P256_SHA256_HKDF_HMAC : public Spake2p
+{
+public:
+    Spake2p_P256_SHA256_HKDF_HMAC(void) : Spake2p(kP256_FE_Length, kP256_Point_Length, kSHA256_Hash_Length) {}
+    virtual ~Spake2p_P256_SHA256_HKDF_HMAC(void) { FreeImpl(); }
```
Comment (dougsteed) [inline src/crypto/CHIPCryptoPAL.h:653]: You're calling a virtual function during a destructor. I think it is actually OK here, but some consider it bad practice (see https://www.artima.com/cppsource/nevercall.html). Maybe there's no reason for FreeImpl to be defined as (pure) virtual in the base class ?

**Pull request merged**